File: personal-assistant-agent/src/agents/llm_coordinator.py
import json
import os
from typing import Any, List

from src.agents.calendar_agent import CalendarAgent
from src.agents.email_agent import EmailAgent
from src.agents.time_agent import TimeAgent
from src.config.llm_config import resolve_llm_settings
from src.services.google_services_utils import GoogleServicesUtils
from src.utils.date_utils import local_iso_date

GOOGLE_SCOPES = [
    "https://www.googleapis.com/auth/gmail.readonly",
    "https://www.googleapis.com/auth/calendar.events.readonly",
]

SYSTEM_PROMPT = """You are a personal assistant with access to the user's Gmail, Google Calendar, and system clock.

Use tools when the user asks about unread email, inbox, messages, calendar, meetings, schedule, or the current time.
Combine multiple tools when a question needs more than one data source (for example, "summarize my morning" may need email and calendar).

When presenting results:
- Be concise and readable.
- For emails, highlight sender and subject.
- For calendar events, show time windows and titles.
- Always call fetch_daily_schedule for calendar questions; never guess.
- fetch_daily_schedule returns { date, timeZone, events }. If events is empty, say no events for that date.
- For "today", omit the date argument so the tool uses the user's local today."""

COORDINATOR_TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "check_unread_emails",
            "description": "Fetch unread emails from the user Gmail inbox.",
            "parameters": {
                "type": "object",
                "properties": {
                    "maxResults": {
                        "type": "number",
                        "description": "Maximum number of emails to return (default 10).",
                    },
                },
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "fetch_daily_schedule",
            "description": "Fetch Google Calendar events for one day in the user local timezone. Omit date for today.",
            "parameters": {
                "type": "object",
                "properties": {
                    "date": {
                        "type": "string",
                        "description": "Optional. Local calendar date YYYY-MM-DD. Leave empty for today.",
                    },
                },
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "get_current_time",
            "description": "Get the current system time in ISO format.",
            "parameters": {"type": "object", "properties": {}},
        },
    },
]


class LlmCoordinator:
    def __init__(self) -> None:
        llm = resolve_llm_settings()
        self.openai = llm.client
        self.model = llm.model
        self.llm_label = f"{llm.label} ({llm.model})"
        self.google = GoogleServicesUtils(
            os.environ.get("CLIENT_ID", ""),
            os.environ.get("CLIENT_SECRET", ""),
            os.environ.get("REDIRECT_URI", ""),
        )
        self.email_agent = EmailAgent(self.google)
        self.calendar_agent = CalendarAgent(self.google)
        self.time_agent = TimeAgent()
        self.history: List[dict[str, Any]] = []

    @property
    def email(self) -> EmailAgent:
        return self.email_agent

    @property
    def calendar(self) -> CalendarAgent:
        return self.calendar_agent

    @property
    def time(self) -> TimeAgent:
        return self.time_agent

    def authenticate(self) -> None:
        self.google.authenticate(GOOGLE_SCOPES)

    def chat(self, user_message: str) -> str:
        self.history.append({"role": "user", "content": user_message})

        while True:
            response = self.openai.chat.completions.create(
                model=self.model,
                messages=[{"role": "system", "content": SYSTEM_PROMPT}] + self.history,
                tools=COORDINATOR_TOOLS,
                tool_choice="auto",
            )

            message = response.choices[0].message
            if not message:
                raise RuntimeError("No response from the language model.")

            self.history.append(message.model_dump(exclude_none=True))

            if not message.tool_calls:
                return (message.content or "(No response)").strip()

            for tool_call in message.tool_calls:
                if tool_call.type != "function":
                    continue
                result = self._run_tool(
                    tool_call.function.name,
                    tool_call.function.arguments or "{}",
                )
                self.history.append(
                    {
                        "role": "tool",
                        "tool_call_id": tool_call.id,
                        "content": json.dumps(result),
                    }
                )

    def _run_tool(self, name: str, args_json: str) -> Any:
        try:
            args = json.loads(args_json) if args_json else {}
        except json.JSONDecodeError as exc:
            raise RuntimeError(f"Invalid tool arguments for {name}") from exc

        if name == "check_unread_emails":
            max_results = args.get("maxResults")
            if not isinstance(max_results, (int, float)):
                max_results = 10
            return self.email_agent.check_unread_emails(int(max_results))

        if name == "fetch_daily_schedule":
            date = args.get("date")
            if isinstance(date, str) and date.strip():
                date_str = date.strip()
            else:
                date_str = local_iso_date()
            schedule = self.calendar_agent.fetch_daily_meeting_schedule(date_str)
            return schedule.to_dict()

        if name == "get_current_time":
            return {"time": self.time_agent.get_current_time()}

        raise RuntimeError(f"Unknown tool: {name}")

File: personal-assistant-agent/src/agents/llmCoordinator.ts
import OpenAI from 'openai';
import { resolveLlmSettings } from '../config/llmConfig';
import { GoogleServicesUtils } from '../services/GoogleServicesUtils';
import { EmailAgent } from './emailAgent';
import { CalendarAgent } from './calendarAgent';
import { TimeAgent } from './timeAgent';
import { localIsoDate } from '../utils/dateUtils';

const GOOGLE_SCOPES = [
    'https://www.googleapis.com/auth/gmail.readonly',
    'https://www.googleapis.com/auth/calendar.events.readonly',
];

const SYSTEM_PROMPT = `You are a personal assistant with access to the user's Gmail, Google Calendar, and system clock.

Use tools when the user asks about unread email, inbox, messages, calendar, meetings, schedule, or the current time.
Combine multiple tools when a question needs more than one data source (for example, "summarize my morning" may need email and calendar).

When presenting results:
- Be concise and readable.
- For emails, highlight sender and subject.
- For calendar events, show time windows and titles.
- Always call fetch_daily_schedule for calendar questions; never guess.
- fetch_daily_schedule returns { date, timeZone, events }. If events is empty, say no events for that date.
- For "today", omit the date argument so the tool uses the user's local today.`;

const COORDINATOR_TOOLS: OpenAI.Chat.Completions.ChatCompletionTool[] = [
    {
        type: 'function',
        function: {
            name: 'check_unread_emails',
            description: 'Fetch unread emails from the user Gmail inbox.',
            parameters: {
                type: 'object',
                properties: {
                    maxResults: {
                        type: 'number',
                        description: 'Maximum number of emails to return (default 10).',
                    },
                },
            },
        },
    },
    {
        type: 'function',
        function: {
            name: 'fetch_daily_schedule',
            description:
                'Fetch Google Calendar events for one day in the user local timezone. Omit date for today.',
            parameters: {
                type: 'object',
                properties: {
                    date: {
                        type: 'string',
                        description:
                            'Optional. Local calendar date YYYY-MM-DD. Leave empty for today.',
                    },
                },
            },
        },
    },
    {
        type: 'function',
        function: {
            name: 'get_current_time',
            description: 'Get the current system time in ISO format.',
            parameters: {
                type: 'object',
                properties: {},
            },
        },
    },
];

export class LlmCoordinator {
    private readonly openai: OpenAI;
    private readonly model: string;
    readonly llmLabel: string;
    private readonly google: GoogleServicesUtils;
    private readonly emailAgent: EmailAgent;
    private readonly calendarAgent: CalendarAgent;
    private readonly timeAgent: TimeAgent;
    private readonly history: OpenAI.Chat.Completions.ChatCompletionMessageParam[] = [];

    constructor() {
        const llm = resolveLlmSettings();
        this.openai = llm.client;
        this.model = llm.model;
        this.llmLabel = `${llm.label} (${llm.model})`;
        this.google = new GoogleServicesUtils(
            process.env.CLIENT_ID || '',
            process.env.CLIENT_SECRET || '',
            process.env.REDIRECT_URI || ''
        );
        this.emailAgent = new EmailAgent(this.google);
        this.calendarAgent = new CalendarAgent(this.google);
        this.timeAgent = new TimeAgent();
    }

    get email() {
        return this.emailAgent;
    }

    get calendar() {
        return this.calendarAgent;
    }

    get time() {
        return this.timeAgent;
    }

    async authenticate(): Promise<void> {
        await this.google.authenticate(GOOGLE_SCOPES);
    }

    async chat(userMessage: string): Promise<string> {
        this.history.push({ role: 'user', content: userMessage });

        while (true) {
            const response = await this.openai.chat.completions.create({
                model: this.model,
                messages: [
                    { role: 'system', content: SYSTEM_PROMPT },
                    ...this.history,
                ],
                tools: COORDINATOR_TOOLS,
                tool_choice: 'auto',
            });

            const assistantMessage = response.choices[0]?.message;
            if (!assistantMessage) {
                throw new Error('No response from the language model.');
            }

            this.history.push(assistantMessage);

            const toolCalls = assistantMessage.tool_calls;
            if (!toolCalls?.length) {
                return assistantMessage.content?.trim() || '(No response)';
            }

            for (const toolCall of toolCalls) {
                if (toolCall.type !== 'function') {
                    continue;
                }

                const result = await this.runTool(
                    toolCall.function.name,
                    toolCall.function.arguments
                );

                this.history.push({
                    role: 'tool',
                    tool_call_id: toolCall.id,
                    content: JSON.stringify(result),
                });
            }
        }
    }

    private async runTool(name: string, argsJson: string): Promise<unknown> {
        let args: Record<string, unknown> = {};
        if (argsJson) {
            try {
                args = JSON.parse(argsJson) as Record<string, unknown>;
            } catch {
                throw new Error(`Invalid tool arguments for ${name}`);
            }
        }

        switch (name) {
            case 'check_unread_emails': {
                const maxResults =
                    typeof args.maxResults === 'number' ? args.maxResults : 10;
                return this.emailAgent.checkUnreadEmails(maxResults);
            }
            case 'fetch_daily_schedule': {
                const date =
                    typeof args.date === 'string' && args.date.trim()
                        ? args.date.trim()
                        : localIsoDate();
                return this.calendarAgent.fetchDailyMeetingSchedule(date);
            }
            case 'get_current_time':
                return { time: await this.timeAgent.getCurrentTime() };
            default:
                throw new Error(`Unknown tool: ${name}`);
        }
    }
}
